### PR TITLE
[5756 Identity Management Overhaul] Injector & daprd 

### PR DESCRIPTION
**Pull request by JoshVanL**
Part of https://github.com/dapr/dapr/issues/5756

Branched from https://github.com/dapr/dapr/pull/6171

Uses security package for identity and serving.

### CODE REVIEW

File: pkg/security/x509source.go [hunk 1]
```diff
@@ -277,6 +316,91 @@ func (x *x509source) requestFromSentry(ctx context.Context, csrDER []byte) ([]*x
...
+	return nil
+}
+
+func (x *x509source) updateTrustAnchorFromFile(filepath string) error {
+	x.lock.RLock()
+	defer x.lock.RUnlock()
+
+	rootPEMs, err := os.ReadFile(filepath)
+	if err != nil {
+		return fmt.Errorf("failed to read trust anchors file %q: %s", filepath, err)
```
Comment (filintod) [inline pkg/security/x509source.go:355]: any reason to not wrap the error `%w`?

File: pkg/security/x509source.go [hunk 2]
```diff
@@ -277,6 +316,91 @@ func (x *x509source) requestFromSentry(ctx context.Context, csrDER []byte) ([]*x
...
+	return nil
+}
+
+func atomicWrite(clock clock.Clock, dir string, data map[string][]byte) error {
+	dir = filepath.Clean(dir)
+	if err := os.MkdirAll(filepath.Dir(dir), 0o700); err != nil {
+		return err
+	}
+
+	newDir := dir + "-" + clock.Now().Format(time.RFC3339)
```
Comment (filintod) [inline pkg/security/x509source.go:382]: maybe not too important in the container, but should we remove the dir on error?

File: cmd/injector/main.go [hunk 3]
```diff
@@ -105,6 +124,43 @@ func main() {
...
+			caBundle, rErr := sec.CurrentTrustAnchors()
+			if rErr != nil {
+				return rErr
+			}
+
+			for {
+				_, rErr = kubeClient.AdmissionregistrationV1().MutatingWebhookConfigurations().Patch(ctx,
+					"dapr-sidecar-injector",
+					types.JSONPatchType,
+					[]byte(`[{"op":"replace","path":"/webhooks/0/clientConfig/caBundle","value":"`+base64.StdEncoding.EncodeToString(caBundle)+`"}]`),
```
Comment (ItalyPaleAle) [inline cmd/injector/main.go:150]: What does this do?

File: pkg/grpc/grpc.go [hunk 4]
```diff
@@ -207,29 +203,7 @@ func (g *Manager) connectRemote(
...
-		ta := credentials.NewTLS(&tls.Config{
-			ServerName:   serverName,
-			Certificates: []tls.Certificate{cert},
-			RootCAs:      signedCert.TrustChain,
-		})
-		opts = append(opts, grpc.WithTransportCredentials(ta))
-	} else {
-		opts = append(opts, grpc.WithTransportCredentials(insecure.NewCredentials()))
-	}
+	opts = append(opts, g.sec.GRPCDialOptionUnknownTrustDomain(namespace, id))
```
Comment (ItalyPaleAle) [inline pkg/grpc/grpc.go:206]: Would this cause issues if a sidecar was trying to connect to a v1.11 sidecar?

File: pkg/injector/patcher/sidecar.go [hunk 5]
```diff
@@ -22,6 +22,7 @@ import (
 	corev1 "k8s.io/api/core/v1"
 
 	injectorConsts "github.com/dapr/dapr/pkg/injector/consts"
+	"github.com/dapr/dapr/pkg/security"
```
Comment (ItalyPaleAle) [inline pkg/injector/patcher/sidecar.go:25]: pkg/injector just went through a refactoring to make it have minimal dependencies on the rest of Dapr. THis is because it can be used as a library outside of Dapr. Looks like pulling in pkg/security brings in a lot, including transitive dependencies such as pkg/diagnostics. What is this package used for in the injector, and can an alternative be found please?

File: pkg/injector/patcher/sidecar_container.go [hunk 6]
```diff
@@ -70,6 +70,9 @@ func (c *SidecarConfig) getSidecarContainer(opts getSidecarContainerOpts) (*core
 		"--app-protocol", c.AppProtocol,
 		"--log-level", c.LogLevel,
 		"--dapr-graceful-shutdown-seconds", strconv.Itoa(c.GracefulShutdownSeconds),
+		"--control-plane-namespace", c.sec.ControlPlaneNamespace(),
```
Comment (ItalyPaleAle) [inline pkg/injector/patcher/sidecar_container.go:73]: This will make the injector unable to inject Dapr sidecars running older versions of Dapr, since they'll see an unsupported CLI flag and crash. The E2E tests will fail with this.

File: pkg/injector/patcher/sidecar_container.go [hunk 7]
```diff
@@ -84,10 +87,6 @@ func (c *SidecarConfig) getSidecarContainer(opts getSidecarContainerOpts) (*core
 		args = append(args, "--control-plane-address", c.OperatorAddress)
 	}
 
-	if c.SentryAddress != "" {
```
Comment (ItalyPaleAle) [inline pkg/injector/patcher/sidecar_container.go:87]: Why was this removed? Can't SentryAddress be empty?

File: charts/dapr/charts/dapr_sidecar_injector/templates/dapr_sidecar_injector_deployment.yaml [hunk 8]
```diff
@@ -191,13 +185,23 @@ spec:
...
           readOnly: true
       volumes:
-      - name: cert
-        secret:
-          secretName: dapr-sidecar-injector-cert
+      - name: dapr-trust-bundle
+        configMap:
+          name: dapr-trust-bundle
+      - name: dapr-identity-token
+        projected:
```
Comment (ItalyPaleAle) [inline charts/dapr/charts/dapr_sidecar_injector/templates/dapr_sidecar_injector_deployment.yaml:199]: Comment here too as a reminder that we must support non-projected volumes too (with a JWKS token)

File: pkg/runtime/config.go [hunk 9]
```diff
@@ -122,7 +123,6 @@ type internalConfig struct {
 	allowedOrigins               string
 	standalone                   configmodes.StandaloneConfig
 	kubernetes                   configmodes.KubernetesConfig
-	mTLSEnabled                  bool
```
Comment (ItalyPaleAle) [inline pkg/runtime/config.go:125]: Please don't just move properties around as it risks causing merge conflicts unnecessarily.

File: pkg/security/security.go [hunk 10]
```diff
@@ -82,32 +102,42 @@ type Options struct {
...
 
+	var source *x509source
 	if opts.MTLSEnabled {
-		if len(opts.TrustAnchors) == 0 {
+		if len(opts.TrustAnchors) > 0 && len(opts.TrustAnchorsFile) > 0 {
+			return nil, errors.New("trust anchors cannot be specified in both TrustAnchors and TrustAnchorsFile")
+		}
+
+		if len(opts.TrustAnchors) == 0 && len(opts.TrustAnchorsFile) == 0 {
 			return nil, errors.New("trust anchors are required")
```
Comment (ItalyPaleAle) [inline pkg/security/security.go:136]:
```suggestion
			return nil, errors.New("trust anchors are required: either one of TrustAnchors or TrustAnchorsFile")
```

File: pkg/security/security.go [hunk 11]
```diff
@@ -140,20 +172,52 @@ func (p *provider) Start(ctx context.Context) error {
...
+				log.Infof("Watching trust anchors file '%s' for changes", p.trustAnchorsFile)
+				return fswatcher.Watch(ctx, p.trustAnchorsFile, caEvent)
+			},
+			func(ctx context.Context) error {
+				for {
+					select {
+					case <-ctx.Done():
+						return nil
+					case <-caEvent:
+						log.Info("trust anchors file changed, reloading trust anchors")
```
Comment (ItalyPaleAle) [inline pkg/security/security.go:202]: Nit:
```suggestion
						log.Info("Trust anchors file changed, reloading trust anchors")
```

File: pkg/security/security.go [hunk 12]
```diff
@@ -167,21 +231,100 @@ func (p *provider) Handler(ctx context.Context) (Handler, error) {
...
+// instruments client authentication using the current signed client
+// certificate. Doesn't verify the servers trust domain, but does authorize the
+// SPIFFE ID path.
+// Used for clients which don't know the servers Trust Domain.
+func (s *security) GRPCDialOptionUnknownTrustDomain(ns, appID string) grpc.DialOption {
+	if !s.mtls {
+		return grpc.WithTransportCredentials(insecure.NewCredentials())
+	}
+
+	expID := fmt.Sprintf("/ns/%s/%s", ns, appID)
```
Comment (ItalyPaleAle) [inline pkg/security/security.go:268]: Nit, since I believe this is used in a hot path:


```suggestion
	expID := "/ns/" + ns + "/" + appID
```

File: pkg/security/security.go [hunk 13]
```diff
@@ -167,21 +231,100 @@ func (p *provider) Handler(ctx context.Context) (Handler, error) {
...
+	s.source.lock.Unlock()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-sub:
+			caBundle, err := s.CurrentTrustAnchors()
+			if err != nil {
+				log.Errorf("failed to marshal trust anchors: %s", err)
```
Comment (ItalyPaleAle) [inline pkg/security/security.go:297]: failed to marshal?

File: pkg/security/x509source.go [hunk 14]
```diff
@@ -84,31 +82,68 @@ type x509source struct {
...
+				return nil, ctx.Err()
+			case <-clock.After(time.Second):
+				log.Warnf("trust anchors file '%s' not found, waiting...", opts.TrustAnchorsFile)
+			}
+		}
+
+		defer f.Close()
+		rootPEMs, err = io.ReadAll(f)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read trust anchors file %q: %w", opts.TrustAnchorsFile, err)
```
Comment (ItalyPaleAle) [inline pkg/security/x509source.go:123]:
```suggestion
			return nil, fmt.Errorf("failed to read trust anchors file '%s': %w", opts.TrustAnchorsFile, err)
```

File: pkg/security/x509source.go [hunk 15]
```diff
@@ -277,6 +316,91 @@ func (x *x509source) requestFromSentry(ctx context.Context, csrDER []byte) ([]*x
...
+
+	if err := atomicWrite(x.clock, *x.writeToDiskDir, map[string][]byte{
+		"tls.crt": certs,
+		"tls.key": pk,
+		"ca.crt":  ta,
+	}); err != nil {
+		return fmt.Errorf("unable to write SVID to disk: %w", err)
+	}
+
+	log.Infof("Wrote SVID to %q", *x.writeToDiskDir)
```
Comment (ItalyPaleAle) [inline pkg/security/x509source.go:344]:
```suggestion
	log.Infof("Wrote SVID to '%s'", *x.writeToDiskDir)
```

File: pkg/security/x509source.go [hunk 16]
```diff
@@ -277,6 +316,91 @@ func (x *x509source) requestFromSentry(ctx context.Context, csrDER []byte) ([]*x
...
+		return err
+	}
+
+	for file, contents := range data {
+		if err := os.WriteFile(filepath.Join(newDir, file), contents, 0o600); err != nil {
+			return err
+		}
+	}
+
+	if err := os.Symlink(newDir, dir+"-new"); err != nil {
```
Comment (ItalyPaleAle) [inline pkg/security/x509source.go:393]: Will this work on Windows without issues?

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]

File: cmd/injector/main.go (hunk 3 above)
Comment (JoshVanL) [reply cmd/injector/main.go:150]: Patches the injector's mutating webhook with the latest trust bundle.

[Number 4 of current shortfalls in the parent issue](https://github.com/dapr/dapr/issues/5756).

File: pkg/grpc/grpc.go (hunk 4 above)
Comment (JoshVanL) [reply pkg/grpc/grpc.go:206]: Yes, we are missing a legacy wrapper for `GRPCDialOptionUnknownTrustDomain`.

File: pkg/injector/patcher/sidecar_container.go (hunk 6 above)
Comment (JoshVanL) [reply pkg/injector/patcher/sidecar_container.go:73]: How have flags been added in the past?

File: pkg/injector/patcher/sidecar_container.go (hunk 7 above)
Comment (JoshVanL) [reply pkg/injector/patcher/sidecar_container.go:87]: No- why would the sentry address ever be empty?

File: pkg/security/security.go (hunk 13 above)
Comment (JoshVanL) [reply pkg/security/security.go:297]: Yes:

```go
// CurrentTrustAnchors returns the current trust anchors for this Dapr
// installation.
func (s *security) CurrentTrustAnchors() ([]byte, error) {
	if s.source == nil {
		return nil, nil
	}
	return s.source.trustAnchors.Marshal()
}
```

File: pkg/security/x509source.go (hunk 16 above)
Comment (JoshVanL) [reply pkg/security/x509source.go:393]: Fixed

File: cmd/injector/main.go (hunk 3 above)
Comment (ItalyPaleAle) [reply cmd/injector/main.go:163]: Got it. Please add a comment explaining what this is doing for future contributors :)

File: pkg/injector/patcher/sidecar_container.go (hunk 7 above)
Comment (ItalyPaleAle) [reply pkg/injector/patcher/sidecar_container.go:87]: It can be empty though, right? In standalone mode daprd runs without sentry. So it's _possible_ even if unlikely.

File: pkg/security/security.go (hunk 13 above)
Comment (ItalyPaleAle) [reply pkg/security/security.go]: Ok. A nit would be to wrap the error in `CurrentTrustAnchors` since the name of the method doesn't imply a connection with marshalling

File: pkg/injector/patcher/sidecar_container.go (hunk 7 above)
Comment (JoshVanL) [reply pkg/injector/patcher/sidecar_container.go:87]: Right, but this injector patching would not be happening in standalone mode.

Comment (ItalyPaleAle) [reply pkg/injector/patcher/sidecar_container.go:87]: Correct. But it's possible to use daprd without sentry. So when someone uses the injector as a library, they have the choice.

File: pkg/injector/patcher/sidecar_container.go (hunk 6 above)
Comment (ItalyPaleAle) [reply pkg/injector/patcher/sidecar_container.go]: This will help for the future, although can't help for today (we need to wait at least 2 more Dapr releases before we can leverage it): https://github.com/dapr/dapr/pull/6766

Comment (JoshVanL) [reply pkg/injector/patcher/sidecar_container.go]: I'm not a fan of ignoring unknown flags. Exiting error on unknown flags is standard for all programs, and makes things super confusing for users to know whether a CLI flag was even picked up. I don't want to think about the human hours wasted by not noticing a spelling mistake in a flag :grimacing: 


We can use environment variables for backwards compat until all supported versions have the CLI flag.

File: pkg/runtime/runtime.go [hunk 17]
```diff
@@ -1459,7 +1457,7 @@ func isEnvVarAllowed(key string) bool {
 	}
 
 	// If we have a `DAPR_ENV_KEYS` env var (which is added by the Dapr Injector in Kubernetes mode), use that as allowlist too
-	allowlist := strings.ToUpper(os.Getenv(securityConsts.EnvKeysEnvVar))
+	allowlist := os.Getenv(securityConsts.EnvKeysEnvVar)
```
Comment (ItalyPaleAle) [inline pkg/runtime/runtime.go:1460]: THis really needs not to be reverted please

```suggestion
	allowlist := strings.ToUpper(os.Getenv(securityConsts.EnvKeysEnvVar))
```

File: pkg/security/consts/consts.go [hunk 18]
```diff
@@ -13,13 +13,15 @@ limitations under the License.
 
 package consts
 
-/* #nosec */
```
Comment (ItalyPaleAle) [inline pkg/security/consts/consts.go:16]: Why was this removed in place of a less ergonomic sequence of `//nolint:gosec`?

File: pkg/security/consts/consts.go [hunk 19]
```diff
@@ -41,9 +43,20 @@ const (
...
 	// SentryTokenFileEnvVar is the environment variable for the Sentry token file.
+	//nolint:gosec
 	SentryTokenFileEnvVar = "DAPR_SENTRY_TOKEN_FILE"
 
 	// AnnotationKeyControlPlane is the annotation to mark a control plane
 	// component. The value is the name of the control plane service.
 	AnnotationKeyControlPlane = "dapr.io/control-plane"
 )
+
+const (
```
Comment (ItalyPaleAle) [inline pkg/security/consts/consts.go:54]:
```suggestion

```

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]

File: pkg/injector/service/config.go [hunk 20]
```diff
@@ -38,6 +36,10 @@ type Config struct {
 	ReadOnlyRootFilesystem            string `envconfig:"SIDECAR_READ_ONLY_ROOT_FILESYSTEM"`
 	SidecarDropALLCapabilities        string `envconfig:"SIDECAR_DROP_ALL_CAPABILITIES"`
 
+	TrustAnchorsFile        string `envconfig:"DAPR_TRUST_ANCHORS_FILE"`
```
[the remark this answers is not in the document]
Comment (JoshVanL) [reply pkg/injector/service/config.go:39]: A filename containing PEM encoded root Certificate Authorities

File: tests/e2e/utils/http.go [hunk 21]
```diff
@@ -41,11 +41,6 @@ func InitHTTPClient(allowHTTP2 bool) {
 	httpClient = NewHTTPClient(allowHTTP2)
 }
```
[the remark this answers is not in the document]
Comment (JoshVanL) [reply tests/e2e/utils/http.go:42]: It's not being used.

[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply tests/e2e/utils/http.go:42]: It was used by a test that was removed, and perhaps could be re-added one day. It still doesn't hurt to keep it.

File: cmd/injector/options/options.go [hunk 22]
```diff
@@ -31,14 +30,16 @@ type Options struct {
...
 	var opts Options
 
 	flag.IntVar(&opts.HealthzPort, "healthz-port", 8080, "The port used for health checks")
 
-	flag.StringVar(&credentials.RootCertFilename, "issuer-ca-secret-key", credentials.RootCertFilename, "Certificate Authority certificate secret key")
-	flag.StringVar(&credentials.IssuerCertFilename, "issuer-certificate-secret-key", credentials.IssuerCertFilename, "Issuer certificate secret key")
-	flag.StringVar(&credentials.IssuerKeyFilename, "issuer-key-secret-key", credentials.IssuerKeyFilename, "Issuer private key secret key")
+	depCAFlag := flag.String("issuer-ca-secret-key", "", "DEPRECATED")
+	depCertFlag := flag.String("issuer-certificate-secret-key", "", "DEPRECATED")
+	depKeyFlag := flag.String("issuer-key-secret-key", "", "DEPRECATED")
```
[the remark this answers is not in the document]
Comment (JoshVanL) [reply cmd/injector/options/options.go]: Done.

File: cmd/injector/options/options.go [hunk 23]
```diff
@@ -31,14 +30,16 @@ type Options struct {
...
+
 func New() *Options {
 	var opts Options
 
 	flag.IntVar(&opts.HealthzPort, "healthz-port", 8080, "The port used for health checks")
 
-	flag.StringVar(&credentials.RootCertFilename, "issuer-ca-secret-key", credentials.RootCertFilename, "Certificate Authority certificate secret key")
-	flag.StringVar(&credentials.IssuerCertFilename, "issuer-certificate-secret-key", credentials.IssuerCertFilename, "Issuer certificate secret key")
-	flag.StringVar(&credentials.IssuerKeyFilename, "issuer-key-secret-key", credentials.IssuerKeyFilename, "Issuer private key secret key")
+	depCAFlag := flag.String("issuer-ca-secret-key", "", "DEPRECATED; Certificate Authority certificate secret key")
```
Comment (ItalyPaleAle) [inline cmd/injector/options/options.go:40]: I think these flags can be deleted right away.

This is a control plane service that is only used in K8s. There's no risk that someone accidentally passes these flags, since the injector is only deployed by Helm.

File: pkg/acl/acl.go [hunk 24]
```diff
@@ -244,16 +151,15 @@ func normalizeOperation(operation string) (string, error) {
 
 func ApplyAccessControlPolicies(ctx context.Context, operation string, httpVerb commonv1pb.HTTPExtension_Verb, isHTTP bool, acl *config.AccessControlList) (bool, string) {
 	// Apply access control list filter
-	spiffeID, err := GetAndParseSpiffeID(ctx)
+	spiffeID, ok, err := spiffe.FromGRPCContext(ctx)
 	if err != nil {
-		// Apply the default action
-		log.Debugf("Error while reading spiffe id from client cert: %v. applying default global policy action", err)
+		log.Debugf(err.Error())
```
Comment (ItalyPaleAle) [inline pkg/acl/acl.go:156]: Leftover log? Or should this at least have some message around it?

File: pkg/acl/acl.go [hunk 25]
```diff
@@ -244,16 +151,15 @@ func normalizeOperation(operation string) (string, error) {
...
 	}
-	var appID, trustDomain, namespace string
-	if spiffeID != nil {
-		appID = spiffeID.AppID
-		namespace = spiffeID.Namespace
-		trustDomain = spiffeID.TrustDomain
+
+	if !ok {
+		// Apply the default action
+		log.Debugf("error while reading spiffe id from client cert. applying default global policy action")
```
Comment (ItalyPaleAle) [inline pkg/acl/acl.go:162]: We uppercase logs
```suggestion
		log.Debugf("Error while reading SPIFFE ID from client cert; applying default global policy action")
```

File: pkg/acl/acl.go [hunk 26]
```diff
@@ -265,37 +171,37 @@ func ApplyAccessControlPolicies(ctx context.Context, operation string, httpVerb
...
 
-	action, actionPolicy := IsOperationAllowedByAccessControlPolicy(spiffeID, appID, operation, httpVerb, isHTTP, acl)
-	emitACLMetrics(actionPolicy, appID, trustDomain, namespace, operation, httpVerb.String(), action)
+	action, actionPolicy := isOperationAllowedByAccessControlPolicy(spiffeID, operation, httpVerb, isHTTP, acl)
+	emitACLMetrics(spiffeID, actionPolicy, operation, httpVerb.String(), action)
 
 	if !action {
-		errMessage = fmt.Sprintf("access control policy has denied access to appid: %s operation: %s verb: %s", appID, operation, httpVerb)
+		errMessage = fmt.Sprintf("access control policy has denied access to id: %s operation: %s verb: %s", spiffeID.URL(), operation, httpVerb)
 		log.Debugf(errMessage)
```
Comment (ItalyPaleAle) [inline pkg/acl/acl.go:179]:
```suggestion
		log.Debug(errMessage)
```

File: pkg/actors/placement/grpc_opts.go [hunk 27]
```diff
@@ -45,17 +46,19 @@ func getGrpcOptsGetter(servers []string, clientCert *daprCredentials.CertChain)
...
 			return nil, errEstablishingTLSConn
 		}
 
+		opts = append(opts, sec.GRPCDialOptionMTLS(placementID))
+
 		if diag.DefaultGRPCMonitoring.IsEnabled() {
 			opts = append(
 				opts,
 				grpc.WithUnaryInterceptor(diag.DefaultGRPCMonitoring.UnaryClientInterceptor()),
-				grpc.WithBlock(),
```
Comment (ItalyPaleAle) [inline pkg/actors/placement/grpc_opts.go:58]: Why was this removed? This is quite important

File: pkg/grpc/server.go [hunk 28]
```diff
@@ -289,87 +264,25 @@ func (s *server) getGRPCServer() (*grpc.Server, error) {
...
-		//nolint:gosec
-		tlsConfig := tls.Config{
-			ClientCAs:  s.signedCert.TrustChain,
-			ClientAuth: tls.RequireAndVerifyClientCert,
-			GetCertificate: func(*tls.ClientHelloInfo) (*tls.Certificate, error) {
-				return &s.tlsCert, nil
-			},
-		}
-
-		// In the internal server, enforce minimum version TLS 1.2
```
Comment (ItalyPaleAle) [inline pkg/grpc/server.go:307]: Looks like this is not enforced anymore in the new code? We need to make sure we are enforcing TLS 1.2+

File: pkg/grpc/server.go [hunk 29]
```diff
@@ -289,87 +264,25 @@ func (s *server) getGRPCServer() (*grpc.Server, error) {
...
+	}
+
 	if s.proxy != nil {
 		opts = append(opts, grpc.UnknownServiceHandler(s.proxy.Handler()))
 	}
 
 	return grpc.NewServer(opts...), nil
 }
 
-func (s *server) startWorkloadCertRotation() {
```
Comment (ItalyPaleAle) [inline pkg/grpc/server.go:335]: Happy to see this gone :) So long, and thanks for all the fish!

File: pkg/injector/service/injector.go [hunk 30]
```diff
@@ -56,12 +55,27 @@ var AllowedServiceAccountInfos = []string{
 	"tekton-pipelines:tekton-pipelines-controller",
 }
 
+type (
+	signDaprdCertificateFn func(context.Context, string) ([]byte, []byte, error)
+	currentTrustAnchorsFn  func() ([]byte, error)
+)
```
Comment (ItalyPaleAle) [inline pkg/injector/service/injector.go:61]: Please, use named arguments and returned value, so we know what this function's expected and returned values are. Just seeing "string" or "[]byte" is not clear

File: pkg/injector/service/injector.go [hunk 31]
```diff
@@ -110,29 +129,28 @@ func getAppIDFromRequest(req *admissionv1.AdmissionRequest) (appID string) {
...
-		config: config,
+		config: opts.Config,
 		deserializer: serializer.NewCodecFactory(
 			runtime.NewScheme(),
 		).UniversalDeserializer(),
 		server: &http.Server{
-			Addr:    fmt.Sprintf(":%d", port),
-			Handler: mux,
-			TLSConfig: &tls.Config{
-				MinVersion: tls.VersionTLS12,
```
Comment (ItalyPaleAle) [inline pkg/injector/service/injector.go:125]: This needs to be kept for all control plane services

File: pkg/security/spiffe/spiffe.go [hunk 32]
```diff
@@ -24,42 +26,77 @@ import (
...
+	id spiffeid.ID
+
+	namespace string
+	appID     string
 }
 
-// FromContext parses a SPIFFE ID from a gRPC context.
-func FromContext(ctx context.Context) (Parsed, bool, error) {
+// FromGRPCContext parses a SPIFFE ID from a gRPC context.
+func FromGRPCContext(ctx context.Context) (*Parsed, bool, error) {
```
Comment (ItalyPaleAle) [inline pkg/security/spiffe/spiffe.go:36]: Question, why was the Parsed returned value changed to a pointer (heap-allocated)?

File: pkg/security/spiffe/spiffe.go [hunk 33]
```diff
@@ -24,42 +26,77 @@ import (
...
 
-	return spiffeid.FromSegments(td, "ns", p.Namespace, p.AppID)
+	return &Parsed{
+		id:        id,
+		namespace: namespace,
+		appID:     appID,
+	}, nil
+}
+
+func (p *Parsed) TrustDomain() spiffeid.TrustDomain {
```
Comment (ItalyPaleAle) [inline pkg/security/spiffe/spiffe.go:76]: All of these methods would probably be not necessary if we did not change From* functions to return a pointer. Which would also cause less allocations on the heap.

File: pkg/acl/acl.go (hunk 24 above)
Comment (JoshVanL) [reply pkg/acl/acl.go:156]: Yeah, the message describing the action has been moved to the log line below. I'll add context to the error here :+1:

File: cmd/injector/options/options.go (hunk 23 above)
Comment (JoshVanL) [reply cmd/injector/options/options.go:40]: I'm happy to do this! I'll make the change.

File: pkg/actors/placement/grpc_opts.go (hunk 27 above)
Comment (JoshVanL) [reply pkg/actors/placement/grpc_opts.go:58]: It is explicitly called out as not recommended in [the docs](https://pkg.go.dev/google.golang.org/grpc#WithBlock), but can add it back if you think we should?

File: pkg/grpc/server.go (hunk 28 above)
Comment (JoshVanL) [reply pkg/grpc/server.go:307]: It is in the spiffe library https://github.com/spiffe/go-spiffe/blob/3000ff09ead100134b2828fec1151b0419a988b0/v2/spiffetls/tlsconfig/config.go#L239

File: pkg/injector/service/injector.go (hunk 31 above)
Comment (JoshVanL) [reply pkg/injector/service/injector.go:125]: Done in the spiffe library.

File: pkg/security/spiffe/spiffe.go (hunk 32 above)
Comment (JoshVanL) [reply pkg/security/spiffe/spiffe.go:36]: I thought this would be preferred as it is being passed around a lot.

File: pkg/security/spiffe/spiffe.go (hunk 33 above)
Comment (JoshVanL) [reply pkg/security/spiffe/spiffe.go:76]: We don't want to make `Parsed` mutable once it has been created is the reason for creating these functions.

File: pkg/actors/placement/grpc_opts.go (hunk 27 above)
Comment (ItalyPaleAle) [reply pkg/actors/placement/grpc_opts.go:58]: It is a fairly significant change actually. We use DialContext to connect to the gRPC server (placement), and the behavior of DialContext [changes](https://pkg.go.dev/google.golang.org/grpc#DialContext) depending on whether WithBlock is set or not:

> In the non-blocking case, the ctx does not act against the connection. It only controls the setup steps.
>
> In the blocking case, ctx can be used to cancel or expire the pending connection. Once this function returns, the cancellation and expiration of ctx will be noop. Users should call ClientConn.Close to terminate all the pending operations after this function returns. 

However, now that I see that... this `grpc.WithBlock` should be added to the options in all cases, and not just if monitoring is enabled (so should be outside of the if block)

**Comment by JoshVanL** [comment]
@ItalyPaleAle Thanks for the review! I have implemented your comments or responded :slightly_smiling_face:

File: pkg/acl/acl.go (hunk 24 above)
Comment (ItalyPaleAle) [reply pkg/acl/acl.go]: Ok. Also change `Debugf` to `Debug` if there's no formatting. Otherwise, `%` chars inside the error message are interpreted as formatting things

File: pkg/security/spiffe/spiffe.go (hunk 33 above)
Comment (ItalyPaleAle) [reply pkg/security/spiffe/spiffe.go:76]: Ok

File: pkg/security/spiffe/spiffe.go (hunk 32 above)
Comment (ItalyPaleAle) [reply pkg/security/spiffe/spiffe.go:36]: Ok

File: pkg/actors/placement/grpc_opts.go (hunk 27 above)
Comment (JoshVanL) [reply pkg/actors/placement/grpc_opts.go:58]: I added `grpc.WithBlock()` to the actor placement client by default.

**Comment by mukundansundar** [comment]
> # Dapr Version Skew test (dapr-sidecar-master - 1.11.3)
> 🔗 **[Link to Action run](https://github.com/dapr/dapr/actions/runs/6132792434)**
> 
> Commit ref: [fb64dfc](https://github.com/dapr/dapr/commit/fb64dfcfe6e51840d9b1c018e2b84423a59e6c29)
> 
> ## ❌ Version Skew tests failed
> Please check the logs for details on the error.

There seems to be a panic seg fault in actor tests in this run.

File: pkg/acl/acl.go [hunk 34]
```diff
@@ -265,37 +171,37 @@ func ApplyAccessControlPolicies(ctx context.Context, operation string, httpVerb
...
-	action, actionPolicy := IsOperationAllowedByAccessControlPolicy(spiffeID, appID, operation, httpVerb, isHTTP, acl)
-	emitACLMetrics(actionPolicy, appID, trustDomain, namespace, operation, httpVerb.String(), action)
+	action, actionPolicy := isOperationAllowedByAccessControlPolicy(spiffeID, operation, httpVerb, isHTTP, acl)
+	emitACLMetrics(spiffeID, actionPolicy, operation, httpVerb.String(), action)
 
 	if !action {
-		errMessage = fmt.Sprintf("access control policy has denied access to appid: %s operation: %s verb: %s", appID, operation, httpVerb)
-		log.Debugf(errMessage)
+		errMessage = fmt.Sprintf("access control policy has denied access to id: %s operation: %s verb: %s", spiffeID.URL(), operation, httpVerb)
+		log.Debug(errMessage)
```
Comment (daixiang0) [inline pkg/acl/acl.go:179]: Why L170 use `log.Debugf` but use `log.Debug` here?

Comment (ItalyPaleAle) [reply pkg/acl/acl.go:179]: `log.Debugf` is like fmt.Sprintf. In this case we are passing the exact message so we must use `log.Debug`.

**Comment by daixiang0** [comment]
@JoshVanL thanks for great job! It is very important for security and I am happy to see it land in 1.12 release.

Please make CI happy.

**Comment by JoshVanL** [comment]
@daixiang0 CI should be more happy now 🙂

**Review by artursouza — APPROVED** [review]
